Add support for sub queries

glrf opened the pull request:
## Summary

This PR introduces sub queries

For data ingestion, sub queries are handled the same as normal queries. The only difference is that sub queries will also be executed when running report for the main query.

Invoice generation will group reports for sub-facts (i.e. facts generated by a sub-query) as sub-items for query line items. That means if we have a query `foo` with a sub-query `sub-foo` and `sub-foo2` and a query `bar`. When generating an invoice we will get the line items

```
items:
- description: foo
  quantity: 4
  ....
  total: 1337
  subItems:
  - description: sub-foo
    quantity: 1
    ....
  - description: sub-foo2
    quantity: 3
    ....
- description: bar
  quantity: 7
  ....
  total: 42

```
Sub items will not show up in the total of the main item.


## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.

<!--
Remove items that do not apply. For completed items, change [ ] to [x].

NOTE: these things are not required to open a PR and can be done afterwards,
while the PR is open.
-->

glrf commented on pkg/invoice/invoice.go:
```diff
@@ -133,10 +151,20 @@ func invoiceForTenant(ctx context.Context, tx *sqlx.Tx, tenant db.Tenant, year i
-		`SELECT queries.description,
+		`SELECT queries.description, queries.id as query_id, queries.parent_id as parent_query_id,
 				SUM(facts.quantity) as quantity, MIN(facts.quantity) as quantitymin, AVG(facts.quantity) as quantityavg, MAX(facts.quantity) as quantitymax,
 				products.unit, products.amount AS pricePerUnit, discounts.discount,
```
```suggestion
				queries.unit, products.amount AS pricePerUnit, discounts.discount,
```
The current implementation always returns the `unit` of the product and not of the query. Is this correct? According to the documented data model products actually shouldn't have a unit..

corvus-ch replied:
There is a unit on the query which is supposed to be used. I have no idea, where the unit on the product comes from. We probably should get rid of `product.uni` and ensure `query.unit` gets used instead.

glrf replied:
I'm not removing the `product.unit` from the database for now, as this is a breaking change.

simu reviewed (approved):
Implementation LGTM overall.

corvus-ch commented on pkg/report/report_test.go:
```diff
@@ -126,6 +141,54 @@ func (s *ReportSuite) TestReport_RunReportCreatesFact() {
+			"INSERT INTO discounts (source,discount,during) VALUES (:source,:discount,:during) RETURNING *", db.Discount{
+				Source:   "bar-product",
+				Discount: 0,
+				During:   infiniteRange(),
```
I miss a test case for when there are more than one instance of a query with different time ranges. Does this also work if the main query and the sub query has a time range not aligned (main: from March to September, sub from August to unbound)?

glrf replied:
I'm not sure what kind of test case you would want for the report. Creating a report is for a single timestamp. An entry either covers the timestamp or it doesn't.  

But you're right we should add more tests for the invoice generation, I just stuck with the current test format. There is actually probably a bug there.

corvus-ch replied:
> I'm not sure what kind of test case you would want for the report. Creating a report is for a single timestamp. An entry either covers the timestamp or it doesn't.

Given a query named "foo" with during set to "unbound to February"
And a query named "foo" with during set to "March to September"
And a query named "foo" with during set to "October to unbound"
And a sub query named "bar" with parent "foo" with during set to "unbound to January"
And a sub query named "bar" with parent "foo" with during set to "January to July"
And a sub query named "bar" with parent "foo" with during set to "August to November"
And a sub query named "bar" with parent "foo" with during set to "December to unbound"

If I run the invoice generation for each month, are the correct queries and sub queries used? This might be already covered in some test, not showing in the diff. I just want to be sure that the time ranges of queries and sub queries can be independently of each other and that this works reliably.

glrf replied:
So yes I agree for invoices. This comment is on a report test (i.e. data ingestion).

And no this is not covered. You're example should work. Changes mid month won't as far as I can tell.
I will probably add end-to-end tests for these scenarios. The current testing approach is just not viable (need >200 lines of code to prepare the db for a single invoice test case)

corvus-ch replied:
As discussed in yesterday's daily, we will cover this in a PR on its own.

corvus-ch reviewed (approved):
Preferably, add some more test cases. I would sleep better, if we test also different cases of time ranges (bound unbound overlapping (main and sub).

corvus-ch reviewed (changes requested):
Sorry, miss clicked. This needs some work.

glrf commented:
I fixed two things:

* Return the query unit for a line item instead of the product unit
* Reproduced and fixed a bug I introduced that dropped line items if there are multiple discounts for one query in a single month.

As discussed I will refactor the tests in a separate PR
